### ISSUE TITLE
Improve documentation for IPython.core.debugger and IPython.terminal.debugger

### ISSUE THREAD

**Issue by tillahoffmann** [ipython/ipython#9940]
The documentation for `IPython.core.debugger` states that it is deprecated but it doesn't provide any information about what should be used instead. A few questions:

* Was `IPython.core.debugger` deprecated in favour of `IPython.terminal.debugger`? 
* If so, what should be used in jupyter notebooks (`IPython.terminal.debugger.set_trace()` doesn't appear to work when used in a notebook)?

**Comment by willingc** [comment]
@tillahoffmann Sorry this isn't clearer in the documentation. Pinging @takluyver and @carreau.

**Comment by takluyver** [comment]
I don't think the module is deprecated - where do you see that? The `Tracer` class is; we should probably add a `set_trace` function to core.debugger as well as terminal.debugger.

**Comment by tillahoffmann** [comment]
Yes, sorry, my description was not specific enough. Is this the correct way to set a breakpoint in the `add` function using the `core` debugger?

```python
from IPython.core.debugger import Pdb
debugger = Pdb()


def add(a, b):
    debugger.set_trace()
    return a + b
```

**Comment by takluyver** [comment]
Not sure, I don't actually use a debugger much ;-). It looks correctish, though, so if it works, go for it!

**Comment by willingc** [comment]
@tillahoffmann And if it works, please let us know and we can open an issue to improve debugger documentation. Thanks.

**Comment by Carreau** [comment]
You should be able to do

```python
from IPython.terminal.debugger import set_trace
set_trace()
```

The difference between `IPython.terminal.debugger` and `IPython.core.debugger` are minimal, 
the former being slightly more aware of the terminal frontend.

**Comment by takluyver** [comment]
I think this arose because `IPython.terminal.debugger` doesn't work in a notebook (or in IPython running as a Jupyter kernel generally). We should probably add `IPython.core.debugger.set_trace()` for such cases.

**Comment by Carreau** [comment]
> I think this arose because IPython.terminal.debugger doesn't work in a notebook (or in IPython running as a Jupyter kernel generally). We should probably add IPython.core.debugger.set_trace() for such cases.

Yep, and we should likely figure out how to nest things a bit less.

**Comment by tillahoffmann** [comment]
> I think this arose because IPython.terminal.debugger doesn't work in a notebook (or in IPython running as a Jupyter kernel generally). We should probably add IPython.core.debugger.set_trace() for such cases.

Yes, that's correct. Thank you all for clarifying the issue. I will use `IPython.core.debugger` and create a similar wrapper as `IPython.terminal.debugger`.

**Comment by TheRealTroff** [comment]
I'm pretty much at my wits end here trying to figure out how this is supposed to work. If I'm using QtConsole and hoping to break into a debugger at some point, what am I supposed to use and how could I have figured that out without asking here? 

```
IPython.terminal.debugger.set_trace() =>  

AttributeError: 'ZMQInteractiveShell' object has no attribute 'editing_mode'
```
while
```
IPython.core.debugger.Tracer()() =>

MultipleInstanceError: Multiple incompatible subclass instances of TerminalIPythonApp are being created.
```

This is running latest everything according to `pip`

**Comment by Carreau** [comment]
The first issue is normal, as `IPython.terminal.debugger` is meant to be used from Terminal IPython only.

`Tracer` is deprecated, use something like : 

```
In [1]: def foo():
   ...:     from IPython.core.debugger import Pdb
   ...:     a =1
   ...:     Pdb().set_trace()
   ...:     
```

In ` IPython.core.debugger` we likely want to define the following function at module level:

```
def set_trace(*args, **kwargs):
    return Pdb().set_trace(*args, **kwargs)
```

Or similar.

**Comment by tillahoffmann** [comment]
I now use the following in my own code to set (conditional) breakpoints.

```python
from IPython.core.debugger import Pdb as CorePdb
import sys


def breakpoint(condition=True):
    """
    Set a breakpoint at the location the function is called if `condition == True`.
    """
    if condition:
        debugger = CorePdb()
        frame = sys._getframe()
        debugger.set_trace(frame.f_back)
        return debugger
    
    
def add(a, b):
    c = a + b
    breakpoint(a + b == 3)
    return c
    
add(1, 2)
```

**Comment by Carreau** [comment]
> Yes, that's correct. Thank you all for clarifying the issue. I will use IPython.core.debugger and create a similar wrapper as IPython.terminal.debugger.

@tillahoffmann do you want to make a PR ? Otherwise @thesageinpilani  is trying to contribute, that would be an easy feature to add !

**Comment by tillahoffmann** [comment]
Happy to submit a PR. Would you prefer the `breakpoint` syntax above or `set_trace` for consistency with `IPython.terminal.debugger`?

**Comment by Carreau** [comment]
I would go with `set_trace()` for consistency with perminal and PDB itself.
